### DISCUSSION TITLE
Custom ViewModel Causes App Crash

### DISCUSSION THREAD

**Discussion by gazishahi** [netreconlab/Parse-Swift#154] [Q&A] [answered]
Hi, I am trying to use ParseSwift with Back4App for my iOS app. I am trying to load some data from Back4App and have it available for use across the entire app, so I thought of using a custom ViewModel. However, simply creating the ViewModel causes the entire app to crash and I get the error 
`ParseSwift/Parse.swift:106: Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value`
It says the error is on the configuration and sessionDelegate. 

When not using the custom ViewModel and only having the data loaded into a single view, I don't run into any errors. Here is my code below.

AppDelegate:
```swift
import UIKit
import ParseSwift

class AppDelegate: UIResponder, UIApplicationDelegate {
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        Task {
            do {
                try await initializeParse()
            } catch {
                print("Parse could not be configured")
            }
        }
            return true
    }
}

private extension AppDelegate {
    func initializeParse() async throws {
        try await ParseSwift.initialize(
            applicationId: "PARSE_APP_ID",
            clientKey: "CLIENT_KEY",
            serverURL: URL(string: "https://parseapi.back4app.com")!)
    }
}
``` 

Custom ViewModel and struct:
```swift
struct relayDictionary: ParseObject {
    var originalData: Data?
    var objectId: String?
    var createdAt: Date?
    var updatedAt: Date?
    var ACL: ParseACL?
    
    var category: String?
    var arabicTerm: String?
    var transliteration: String?
    var litDef: String?
    var techDef: String?
}

extension relayDictionary {
    init(category: String, arabicTerm: String, transliteration: String, litDef: String, techDef: String) {
        self.category = ""
        self.arabicTerm = ""
        self.transliteration = ""
        self.litDef = ""
        self.techDef = ""
    }
}

class ViewModel: ObservableObject {
    @Published var objects = [relayDictionary]()
    @Published var error = ParseError?(.init(code: .invalidQuery, message: "Invalid query"))
    
    private var subscriptions = Set<AnyCancellable>()
    
    init() {
        fetchDict()
    }
    
    func fetchDict() {
        let query = relayDictionary.query("category" == "greetings")
        let publisher = query
            .findPublisher()
            .sink(receiveCompletion: { result in
                switch result {
                case .failure(let error):
                    //: Publish error.
                    self.error = error
                    case .finished:
                    print("Successfully queried data")
                    }
                },
                receiveValue: {
                    //: Publish found objects
                    self.objects = $0
                    print("Found \(self.objects.count), objects: \(self.objects)")
                })
        publisher.store(in: &subscriptions)
    }
}
``` 

In my ContentView, I have
`@StateObject var dictionaryViewModel = ViewModel()`

This line seems to cause the error. Removing it causes my app to function again. I am trying to pass dictionaryViewModel as an @EnvironmentObject in other views, but the app doesn't work at all with that line in ContentView. I don't know if initializeParse() is being skipped because the error I get seems to indicate the URL is nil if I'm understanding it correctly. I'm not exactly sure what the cause here is and what the fix here would be. I've been trying to follow the Playground examples but to no avail. Any help would be greatly appreciated! 

UPDATE:
First, I have resolved the crashes. For some reason, it was the query "category" == "greetings" that caused the crashes. Removing it and just querying the entire class stops the crashes.

I am now having another issue in my code:
In the ViewModel, ParseError is given an error code. I can't just have `@Published var error = ParseError?` without the .init() or else it gives me the error `Expected member name or constructor call after type name`. 

The causes another issue when loading data, as it appears to always force an error, so I can't load my data while having 
```swift
if let error = viewModel.error {
                Text(error.description)
            }
``` 
in any part of my code. I get the same hardcoded error as I have in the ViewModel. Removing the if statement remedies the issue but of course I need error handling so I'm not sure how to go about this now.

Is there a way to ensure that my var error in the ViewModel is declared as ParseError? without the .init()?

**Comment by jaysonng** [comment]
Have you tried initializing the ViewModel inside the init() of your view instead of directly initializing it like `@StateObject var dictionaryViewModel = ViewModel()` ?

i.e.

```swift
init() {
     _dictionaryViewModel = StateObject(wrappedValue: ViewModel())
}
```

**Reply by gazishahi** [reply]
I have not tried that, but instead found the query to be the culprit of my crashes. For some reason, having the query as `"category" == "reasons` was causing the crashes, so removing that fixed my issue. I now have another issue about my var error in the ViewModel, I updated my original post to give the details on that.

**Reply by jaysonng** [reply]
just an observation as well. This init you have is wrong. you're not assigning the variables.

You need to remove the  ` = " "` and change it to:

```swift
extension relayDictionary {
    init(category: String = "", arabicTerm: String = "", transliteration: String = "", litDef: String = "", techDef: String = "") {
        self.category = category
        self.arabicTerm = arabicTerm
        self.transliteration = translieteration
        self.litDef = litDef
        self.techDef = techDef
    }
}
```

This basically assigns a blank String to each variable if you don't pass it on init.

for example

```swift
instance.init(category: "New Category") 
```
will initialize it with "category"  as "New Category" and everything else as blank.

**Reply by gazishahi** [reply]
It wasn't giving an error so I assumed it was okay, thank you for the help!

**Comment by cbaker6** [comment]
What version of Parse-Swift are you using? I suspect <5.0.1. You shouldn’t see the issue in >5.0.1 as described in more detail here: https://github.com/netreconlab/Parse-Swift/discussions/37#discussioncomment-5260675

The issue is your code is making calls to the SDK before it finishes initialization which can happen if you are using the old AppDelegate life-cycle with SwiftUI. If you are not ready to migrate to a newer version of Parse-Swift, see workarounds: https://github.com/netreconlab/Parse-Swift/discussions/37#discussioncomment-4742953

**Reply by gazishahi** [reply]
I'm using 5.8.2. The issue was with the query I had, removing that and just querying all the data seems to work fine. My issue now is with ParseError. I'm trying to follow the Playgrounds example [here](https://github.com/netreconlab/Parse-Swift/blob/main/ParseSwift.playground/Pages/18%20-%20SwiftUI%20-%20Finding%20Objects%20With%20Custom%20ViewModel.xcplaygroundpage/Contents.swift), but I can't seem to have ParseError? without using .init(). I updated the post to highlight this issue.

**Reply by cbaker6** [reply, marked as answer]
The Swift Playgrounds is runnable, what happens when you deploy a local server following: https://github.com/netreconlab/Parse-Swift?tab=readme-ov-file#test-drive-parse-swift

and run the actual Playground file which has an optional ParseError: https://github.com/netreconlab/Parse-Swift/blob/b3169f2b438df9bb1d14935246be0a462c97c42f/ParseSwift.playground/Pages/18%20-%20SwiftUI%20-%20Finding%20Objects%20With%20Custom%20ViewModel.xcplaygroundpage/Contents.swift#L76-L108

if the playground file runs correctly, it means you are doing something wrong in your code.

**Reply by gazishahi** [reply]
I was able to get the Playground to run, and yes you're right it was my code. I made the dumbest mistake possible.

I had `@Published var error = ParseError?` instead of `@Published var error: ParseError?`. I was severely overthinking this haha. Thank you for the help!
